### PR TITLE
HPCC-17450 Thor leaks globally-cached MySQL connections

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
Ran the test program in the Jira, and observed unclosed connections. Ran again with the fix and observed them closed. It's not easy to automate this test.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review

File: plugins/mysql/mysqlembed.cpp [hunk 1]
```diff
@@ -427,7 +427,7 @@ class MySQLConnection : public CInterface
         ForEachItemInRev(idx, globalCachedConnections)
         {
             MySQLConnection &cached = globalCachedConnections.item(idx);
-            if (now - cached.created > maxAge)
+            if (!maxAge || (now - cached.created > maxAge))
```
Comment (ghalliday) [inline plugins/mysql/mysqlembed.cpp]: @richardkchapman should this be !maxAge && ?  Otherwise the test is redundant.

Comment (richardkchapman) [inline plugins/mysql/mysqlembed.cpp]: || is right I think

The test is not quite redundant in that now==cached.created is possible.

**Comment by richardkchapman** [comment]
@ghalliday Back to you

File: plugins/mysql/mysqlembed.cpp (hunk 1 above)
Comment (ghalliday) [inline plugins/mysql/mysqlembed.cpp]: I had missed the special case use of retireCache(0) in your new code.

**Pull request merged**